GPII-3255

sgithens opened the pull request:
This is my first draft at factoring out some of the resuable code to be used from universal, the PPT, and potentiall other locations.

sgithens committed:
GPII-3255 Decoupling create user and unlock users from public http endpoints.

the-t-in-rtf commented:
Hi, @sgithens.  Thanks for this.  I will look through it in more depth shortly.  I wanted to ask you up front to:

1. Add docs for the new code-based usage.
2. Update the `README.md` file with a summary of how to use the "utils" from custom code, linking to the above.
3. Update `docs/api.md`  to document any new API endpoints .

In addition to helping users of the library, the docs will help with the review.

the-t-in-rtf commented on src/js/server/api.js:
```diff
@@ -10,6 +10,7 @@
 /* eslint-env node */
 "use strict";
 var fluid = require("infusion");
+var request = require("request"); // TODO:  Replace this with a writable data source.
```
Linting error:

> 13:5  error  'request' is defined but never used  no-unused-vars

the-t-in-rtf commented on src/js/server/api.js:
```diff
@@ -20,7 +21,9 @@ require("./logout.js");
 require("./verify.js");
+require("./utils.js");
 
+var gpii = fluid.registerNamespace("gpii");
```
Linting error:

> 26:5  error  'gpii' is defined but never used     no-unused-vars

the-t-in-rtf commented on src/js/server/login.js:
```diff
@@ -13,25 +13,17 @@ require("./lib/password");
-        if (encodedPassword === response.derived_key) {
-            // Transform the raw response to ensure that nothing sensitive is exposed to the user
-            var user = fluid.model.transformWithRules(response, that.options.rules.user);
+gpii.express.user.login.post.handler.verifyPassword = function (that, utils, request, response) {
```
> 16:87  error  'response' is defined but never used  no-unused-vars

Another spot to clean up after the refactor.

the-t-in-rtf commented:
I will stop leaving individual comments, but obviously `grunt lint` needs to pass before this can be merged.  At the moment it nicely illustrates bits of cruft left behind from the refactor.

the-t-in-rtf commented:
Coverage looks good compared to the branch you're merging with.

the-t-in-rtf commented on src/js/server/login.js:
```diff
@@ -50,8 +42,8 @@ fluid.defaults("gpii.express.user.login.post.handler", {
-            func: "{reader}.get",
-            args: ["{that}.options.request.body"]
+            funcName: "gpii.express.user.login.post.handler.verifyPassword",
+            args: ["{that}", "{gpii.express.user.utils}", "{that}.options.request", "{that}.options.response"]
```
This should also be updated, you're relying on an invoker rather than passing `response` now.

the-t-in-rtf commented on src/js/server/signup.js:
```diff
@@ -19,48 +19,22 @@ gpii.express.user.signup.post.handler.lookupExistingUser = function (that) {
 };
 
-gpii.express.user.signup.post.handler.checkForExistingUser = function (that, response) {
+gpii.express.user.signup.post.handler.checkForExistingUser = function (that, utils, response) {
```
Same deal here, you no longer need to pass the response.

the-t-in-rtf commented on src/js/server/signup.js:
```diff
@@ -108,7 +71,7 @@ fluid.defaults("gpii.express.user.signup.post.handler", {
         "checkForExistingUser": {
             funcName: "gpii.express.user.signup.post.handler.checkForExistingUser",
-            args:     ["{that}", "{arguments}.0"]
+            args:     ["{that}", "{gpii.express.user.utils}", "{arguments}.0"]
```
`{arguments}.0` is no longer needed here.

sgithens replied:
I believe this one is still necessary as it's doing a check for the existence of the username before moving forward.

the-t-in-rtf commented on src/js/server/utils.js:
```diff
@@ -0,0 +1,140 @@
+    }
+});
+
+gpii.express.user.utils.createNewUser = function (that, userData) {
```
If devs are now expected to be able to call these functions directly without using the API, it seems like it's time to add simple JSDocs.

the-t-in-rtf replied:
Should be "user's" throughout.

the-t-in-rtf commented:
OK, I've left comments for everything you changed.  In addition to adding docs for the new features, I'd also like to ask that you add at least simple tests of the new functions in isolation.  In other words, if you intend for devs to be able to use the utils outside of an express instance, you should test that separately to confirm that there's nothing that depends on having express-isms like a response, et cetera.

sgithens committed:
GPII-3255 Minor lint cleanup

sgithens committed:
GPII-3255 JSDocs for refactored routines.

sgithens committed:
GPII-3255 Adding markdown doc for utils component.

the-t-in-rtf commented on src/js/server/utils.js:
```diff
@@ -17,6 +17,15 @@ var request = require("request"); // TODO:  Replace this with a writable data so
 var gpii = fluid.registerNamespace("gpii");
 fluid.registerNamespace("gpii.express.user.utils");
 
+/**
```
I tend to be a bit nervous using JSDocs-style block comments for things other than functions, as someone can easily trip a linting error by adding a function.  However, I wanted to ask about your docs creation work and how you would make use of isolated blocks like this there.

the-t-in-rtf commented on src/js/server/utils.js:
```diff
@@ -69,6 +78,21 @@ fluid.defaults("gpii.express.user.utils", {
+ * @param {String} userData.username - New users username
+ * @param {String} userData.email - New users email
+ * @param {String} userData.password - Password for new account
+ * @return {Promise} Promise container either the new Couch Record for
```
containing?

the-t-in-rtf replied:
I'd suggest lowercasing "record" here.  Also, I think "CouchDB" is probably better than shortening to "Couch".

the-t-in-rtf commented on src/js/server/utils.js:
```diff
@@ -108,12 +132,36 @@ gpii.express.user.utils.createNewUser = function (that, userData) {
+ * @param {gpii.express.user.utils} that - Utils component.
+ * @param {String} username - Username to use for record lookup.
+ * @param {String} password - Clear text password to validate record with.
+ * @return {Object} The userData record if the password is correct, otherwise
```
either add backticks around "userData" or break up into the words, i.e. "user data".

the-t-in-rtf commented on src/js/server/utils.js:
```diff
@@ -108,12 +132,36 @@ gpii.express.user.utils.createNewUser = function (that, userData) {
+ *
+ * Attempts to look up username using the current CouchDB view. (At the time
+ * of writing either username or email). And unlock their account using `password`.
+ * If the username and password match, the Couch userData record will be returned.
```
Here the use of "Couch" seems fine to me, as the full name appears nearby.

the-t-in-rtf commented on src/js/server/utils.js:
```diff
@@ -108,12 +132,36 @@ gpii.express.user.utils.createNewUser = function (that, userData) {
 };
 
+/**
+ * gpii.express.user.utils.verifyPassword function
```
What value do these add versus the code itself?  Are they useful in your docs generation?

the-t-in-rtf commented:
@sgithens, I left comments on the JSDocs.  We still need README and docs changes, I assume you're working on those.

sgithens committed:
GPII-3255 README work based on PR Review

sgithens committed:
GPII-3255 JSDOC work based on PR Review

sgithens committed:
GPII-3255 In progress unit tests

sgithens commented:
@the-t-in-rtf I think I've addressed most of the above, with the exception of a few things I need you to look at, but I'll recreate this pull against GPII master and then put those notes in.

sgithens committed:
GPII-3255 Some cleanup of unit tests and doco

sgithens committed:
GPII-3255 Fixing test sequencing

- Added a sequency delay event to fire from a promise based
  test, to stop progress of the sequencing until it completes.

sgithens committed:
GPII-3255 Added remaining tests for utils

- Tests for failure of unlocking with a bad username/password and
  creating a new account.

sgithens committed:
GPII-3255 Replaced delay component with IoC Test Task

sgithens committed:
GPII-3255 Doc and tests updates based on PR

sgithens committed:
GPII-3255 Minor doco changes.

amb26 commented on src/js/server/utils.js:
```diff
@@ -0,0 +1,188 @@
+ *  Expects to be distributed the same couch options as the other components
+ *  in gpii-express-user.
+ */
+fluid.defaults("gpii.express.user.utils", {
```
Utils is far too generic and uninformative a name. Given this component appears to deal mainly with passwords and their encoding I suggest something like "crypto" or "passwords"

amb26 commented on tests/js/server/utils-tests.js:
```diff
@@ -0,0 +1,119 @@
+var fluid        = require("infusion");
+var gpii         = fluid.registerNamespace("gpii");
+
+require("../../../");
```
Use a meaningful import such as fluid.require("%gpii-express-user");

amb26 commented on tests/js/server/utils-tests.js:
```diff
@@ -0,0 +1,119 @@
+/* eslint-env node */
+"use strict";
+
+var fluid        = require("infusion");
```
Bizarre massive spaces here

sgithens commented:
Closing this pull request, as it's been merged in previously in another branch.
